In [28]:
import numpy as np
import pandas as pd
from scipy import stats

In [35]:
def mode(Series):
    res = pd.DataFrame()
    res["counts"] = Series.value_counts()
    ret = res.loc[res["counts"]==res.counts.max()]
    
    if len(ret) < len(res):
        return pd.Series([x for x in res.index])
    elif len(ret)==len(res):
        return("No mode")

In [36]:
df = pd.DataFrame()
df["ages"] = [14, 12, 11, 10, 8, 6]

In [37]:
df.describe()

ages
count   6.000000
mean   10.166667
std     2.857738
min     6.000000
25%     8.500000
50%    10.500000
75%    11.750000
max    14.000000

## Question 1: Descriptive statistics calculated by hand from initial values

- Mean = 10.167
- Median = 10.5
- Mode = None, all values were completely unique
- Variance = 8.167
- Standard deviation = 2.86
- Standard error = 1.17

```python
df.description()
```

In [43]:
# Confirming hand calculations with Pandas
# I used the numpy standard deviation function instead of using the describe method
# to get the sample standard deviation rather than the population standard deviation.

def descriptive(series):
    results = pd.DataFrame()
    results["values"] = [np.var(series, ddof=1),
                        np.std(series, ddof=1),
                        stats.sem(series, ddof=1),
                        np.mean(series),
                        np.median(series),
                        mode(series)]
    results.index = ["Variance", "Standard deviation", "Standard error", "Mean", "Median", "Mode"]
    return(results)

descriptive(df["ages"])

values
Variance            8.16667
Standard deviation  2.85774
Standard error      1.16667
Mean                10.1667
Median                 10.5
Mode                No mode

## Question 2
If I could only use one cenral tendency I would use median because it's less likely to be influenced by outliers.

## Question 3
__Question:__ Next, Cindy has a birthday. Update your estimates- what changed, and what didn't?
<br><br>
__Answer:__ The mean will slightly increase, as will the variance and standard error. The median and mode will stay the same. So far standard error seems least likely to be impacted by extreme variables so I would probably use that and median for the variation and central tendency measures.

## Question 4
__Question:__ Replace Cousin Oliver with 1-year-old Jessica, then recalculate again. Does this change your choice of central tendency or variance estimation methods?
<br><br>
__Answer:__ I would choose the median and standard error. Standard error has the strongest relationship to the sample size and therefore not outliers. I choose median for the same reason. With a larger sample size these wouldn't matter as much, though.

In [48]:
oliver = pd.DataFrame([14, 12, 11, 10, 6, 1], columns=["ages"])
descriptive(oliver["ages"])

values
Variance               22.4
Standard deviation  4.73286
Standard error      1.93218
Mean                      9
Median                 10.5
Mode                No mode

## Question 5
__Question:__ Based on these numbers, what percentage of adult Americans would you estimate were Brady Bunch fans on the 50th anniversary of the show?
<br><br>
__Answer:__ Based on the very small sample size I would use the median to estimate the viewership, which in this case is calculated below.

In [49]:
np.median([0.23, 0.2, 0.17, 0.05])

0.185